### ISSUE TITLE
Incorrect Gradient Computation in `torch.log1p`

### ISSUE THREAD

**Issue by vwrewsge** [pytorch/pytorch#152088]
### 🐛 Describe the bug

# To Reproduce
```python
import torch
def test_bug():
    a = torch.tensor([-1.0, 0.5, 1.0], requires_grad=True)
    l = torch.log1p(a)[a > -1].sum()  # This will include only a[1] and a[2]
    l.backward()
    print(a.grad)
if __name__ == "__main__":
    test_bug()
```

# Output
```
tensor([   nan, 0.6667, 0.5000])
```

# Expected Behaviour
Since `a[0] = -1.0` is excluded by the `mask (a > -1)`, it should not contribute to the output `l`. Therefore, its gradient (`a.grad[0]`) should be `0` instead of `nan`.

### Versions

PyTorch version: 2.6.0+cu124
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

cc @ezyang @albanD @gqchen @nikitaved @soulitzer @Varal7 @xmfan

**Comment by malfet** [comment]
Adding a triage review, because there doesn't seem to be `module: operators` label (and this doesn't fall under `module: nn`)

**Comment by vadimkantorov** [comment]
I guess, this a good old issue on "torch.where propagating nan gradients even when masked out", but for `masked_select`?
